### PR TITLE
[GRPC] Implement Hybrid Queries

### PR DESCRIPTION
**Pull request by karenyrx**
### Description
Implement GRPC Hybrid Query 

### Test Plan
Built and deployed to internal cluster using this change from opensearch core: https://github.com/opensearch-project/OpenSearch/pull/20103

Sample hybrid query: 
```
{
  "search_request_body": {
    "query": {
      "hybrid": {
        "queries": [
          {
            "match": {
              "field": "test",
              "query": {
                "string": "hi"
              }
            }
          },
          {
            "match_all": {}
          }
        ]
      }
    }
  }
}
```
Response was successful:
```
{
  "body": {
    "took": "88",
    "xShards": {
      "successful": 1,
      "total": 1,
      "skipped": 0
    },
    "hits": {
      "total": {
        "totalHits": {
          "relation": "TOTAL_HITS_RELATION_EQ",
          "value": "1"
        }
      },
      "hits": [
        {
          "xIndex": ".plugins-ml-config",
          "xId": "master_key",
          "xScore": {
            "double": 1
          },
          "xSource": "eyJtYXN0ZXJfa2V5IjoiTEZadjNCWG83SHVjL2U4T2NXTGZoN2czV3U1TDFNVDBjRmMyZkJYSG1GQT0iLCJjcmVhdGVfdGltZSI6MTc2MzQxMTMyODg4OX0="
        }
      ],
      "maxScore": {
        "float": 1
      }
    }
  }
}

```
### Related Issues
Partially resolves https://github.com/opensearch-project/neural-search/issues/1495 

### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by owaiskazi19** [comment]
@karenyrx thanks for integrating GRPC support. Are you planning to add the same for other queries as well in this repo like `neural`, `neural_sparse`, `agentic`?

File: build.gradle [hunk 1]
```diff
@@ -246,7 +246,7 @@ opensearchplugin {
     classname "${projectPath}.${pathToPlugin}.${pluginClassName}"
     licenseFile rootProject.file('LICENSE')
     noticeFile rootProject.file('NOTICE')
-    extendedPlugins = ['opensearch-knn']
+    extendedPlugins = ['opensearch-knn', 'transport-grpc']
```
Comment (karenyrx) [inline build.gradle:250]: Problematic due to https://github.com/opensearch-project/OpenSearch/issues/20100

**Comment by karenyrx** [comment]
hi @owaiskazi19 , after merging https://github.com/opensearch-project/OpenSearch/pull/20103/files, i would expect the error to go away but I'm still seeing this on the CI:
```
Error: Exception in thread "main" java.lang.IllegalStateException: failed to load plugin opensearch-neural-search due to jar hell
	at org.opensearch.plugins.PluginsService.checkBundleJarHell(PluginsService.java:763)
	at org.opensearch.plugins.PluginsService.checkJarHellForPlugin(PluginsService.java:404)
	at org.opensearch.tools.cli.plugin.InstallPluginCommand.jarHellCheck(InstallPluginCommand.java:834)
	at org.opensearch.tools.cli.plugin.InstallPluginCommand.loadPluginInfo(InstallPluginCommand.java:811)
	at org.opensearch.tools.cli.plugin.InstallPluginCommand.installPlugin(InstallPluginCommand.java:846)
	at o
```
Is there a way to check if the CI is using the latest OS core main snapshot or not?

**Comment by vibrantvarun** [comment]
@karenyrx Check this PR https://github.com/opensearch-project/k-NN/pull/3050/changes

Similarly you can also fix the jar hell in this PR 

Do following steps.

1. Create a fresh PR after rebasing with latest main.
2. If issue still persists then see we can resolve this jarhell by referring to the PR mentioned above.

**Comment by vibrantvarun** [comment]
@karenyrx I think your rebase is not done properly as we have merged few PR's in last few days and have not seen issue like this.
Could you create a fresh PR.

**Comment by karenyrx** [comment]
Thanks for checking @vibrantvarun . 

Double checked my rebase has the latest commit on main: 
```
~/neural-search on [grpchybrid] % git log 
commit 188706032b0095cc407de77243fd71be696ea0dd (HEAD -> grpchybrid, upstream/grpchybrid)
Author: Karen X <karenxyr@gmail.com>
Date:   Thu Jan 8 05:12:11 2026 +0000

    [GRPC] Add GRPC support for Hybrid queries
    
    Signed-off-by: Karen X <karenxyr@gmail.com>

commit 5cfe11844652d94d4cf7d8f22e4ebcb961a30522 (ossupstream/main, main)
Author: Zirui Song <zrsong@amazon.com>
Date:   Thu Jan 8 10:45:59 2026 +0800

    [Feature Enhancement] Enable explain function within Sparse ANN query (#1694)
    
    * Enable explain function within Sparse ANN query
    
    Signed-off-by: Zirui Song <zrsong@amazon.com>
```
I also tested an empty commit on this fresh PR https://github.com/opensearch-project/neural-search/pull/1720, and the CI fails with the [same error](https://github.com/opensearch-project/neural-search/actions/runs/20806442829/job/59761659168?pr=1720)
```

Execution failed for task ':compileTestJava'.
> Compilation failed; see the compiler output below.
  /home/runner/work/neural-search/neural-search/src/test/java/org/opensearch/neuralsearch/sparse/SparseBaseIT.java:320: error: no suitable method found for search(String,QueryBuilder,<null>,int,Map<String,String>)
          return search(index, queryBuilder, null, resultSize, Map.of("explain", "true"));
                 ^
      method BaseNeuralSearchIT.search(String,QueryBuilder,int) is not applicable
        (actual and formal argument lists differ in length)
      method BaseNeuralSearchIT.search(String,QueryBuilder,QueryBuilder,int) is not applicable
        (actual and formal argument lists differ in length)
      method BaseNeuralSearchIT.search(String,QueryBuilder,QueryBuilder,i
```

**Comment by vibrantvarun** [comment]
@karenyrx got it. I reached out to @zirui-song-18 to fix the build. Should be fixed soon. Thanks for finding out.

**Comment by karenyrx** [comment]
Thanks @martin-gaievski , while the CI is being fixed, could you or a maintainer help review the rest of this PR? Thanks!

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        this.registry = registry;
+    }
+
+    @Override
+    public QueryContainer.QueryContainerCase getHandledQueryCase() {
+        return QueryContainer.QueryContainerCase.HYBRID;
+    }
+
+    @Override
+    public QueryBuilder fromProto(QueryContainer queryContainer) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java:32]: In cases like hybrid query with filter, some cases of collapse HQ comes wrapped under boolean query. How do we handle that?

File: src/test/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverterTests.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+public class HybridQueryBuilderProtoConverterTests extends OpenSearchTestCase {
+
+    private HybridQueryBuilderProtoConverter converter;
+
+    @Mock
+    private QueryBuilderProtoConverterRegistry mockRegistry;
+
+    @Before
+    public void setup() {
```
Comment (vibrantvarun) [inline src/test/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverterTests.java:42]: Can you add integ test as well?

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java (hunk 2 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java:32]: You mean a query like `hybrid : { filter  : { bool : { collapse : {... }} }}`? Yes that is handled.

The `QueryBuilderProtoConverterRegistry registry`   is accessible to all classes which implement the `QueryBuilderProtoConverter` interface (e.g. `HybridQueryBuilderProtoConverter`) , and contains the query converters for all the query types implemented in the [core transport-grpc module](https://github.com/opensearch-project/OpenSearch/tree/main/modules/transport-grpc/src/main/java/org/opensearch/transport/grpc/proto/request/search/query) (which includes `bool` query), as well as all the query types implemented in plugins which extend the transport-grpc SPI (e.g. `knn` queries from the k-NN plugin, and `hybrid` queries from the neural-search plugin).  

More details in this README: https://github.com/opensearch-project/OpenSearch/pull/18949/files#diff-7f108acb069d1b854fc3758961f8e0d69d403cf3f748e969f4cfbbc80726ab33

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * Converts a Protocol Buffer HybridQuery to an OpenSearch HybridQueryBuilder.
+     * This method follows the exact same pattern as {@link HybridQueryBuilder#fromXContent(XContentParser)}
+     * to ensure parsing consistency and compatibility with the REST API side.
+     *
+     * @param hybridQueryProto The Protocol Buffer HybridQuery to convert
+     * @param registry The registry to use for converting nested queries
+     * @return A configured HybridQueryBuilder instance
+     */
+    public QueryBuilder fromProto(HybridQuery hybridQueryProto, QueryBuilderProtoConverterRegistry registry) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:41]: In current version of the Builder.fromXContent() we track query stats, but here that part is missing. Can you please check existing builder logic and add missing method: https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java#L295

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        if (hybridQueryBuilder.queries().isEmpty()) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "[%s] requires 'queries' field with at least one clause", HybridQueryBuilder.NAME)
+            );
+        }
+
+        if (hybridQueryProto.hasFilter()) {
+            QueryBuilder filterBuilder = registry.fromProto(hybridQueryProto.getFilter());
+            if (filterBuilder != null) {
+                hybridQueryBuilder.filter(filterBuilder);
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java]: this approach is different from existing builder code. you are adding filter only at top level, while in existing build it's added for each sub-query separately https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java#L278
If I remember correctly hybrid query supports both syntaxes - top level and sub-query level https://github.com/opensearch-project/neural-search/issues/1135

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+
+        if (hybridQueryProto.hasPaginationDepth()) {
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java]: here version check is missing, this can be a problem when domain is migrating from lower versions 
```
if (isClusterOnOrAfterMinReqVersionForPaginationInHybridQuery()) {
```

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 5 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java]: +1.

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 4 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:41]: thank you, missed that, added a call to `updateQueryStats()`

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 6 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java]: true, though transport-grpc is only supported on 3.x+ but that is still the correct call to add

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 5 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java]: updated

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        if (queryContainer == null || queryContainer.getQueryContainerCase() != QueryContainer.QueryContainerCase.HYBRID) {
+            throw new IllegalArgumentException("QueryContainer does not contain a Hybrid query");
+        }
+
+        return HybridQueryBuilderProtoUtils.fromProto(queryContainer.getHybrid(), registry);
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java:40]: please add defensive check for registry being not null. this should avoid NPE deep in util class (poor debugging context) in case setRegistry() is never called

**Review by martin-gaievski — COMMENTED** [review]
Overall code looks good, please address my last minor comment and I'm good to approve PR.

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java (hunk 7 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverter.java:40]: thanks good call, done

**Review by martin-gaievski — APPROVED** [review]
Looks good to me, thank you for this contribution and addressing all review comments

File: src/test/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverterTests.java (hunk 3 above)
Comment (karenyrx) [reply src/test/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoConverterTests.java:42]: Good point. Currently the integration test framework for gRPC is still under development. Could we add it in a separate PR once it's ready? Filed an issue to track https://github.com/opensearch-project/neural-search/issues/1723

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            if (queries.size() == HybridQueryBuilder.MAX_NUMBER_OF_SUB_QUERIES) {
+                throw new IllegalArgumentException(
+                    String.format(Locale.ROOT, HybridQueryBuilder.ERROR_MSG_MAX_QUERIES_EXCEEDED, HybridQueryBuilder.NAME)
+                );
+            }
+            QueryBuilder queryBuilder = parseInnerQueryBuilder(queryContainer, registry);
+            if (queryBuilder != null) {
+                queries.add(queryBuilder);
+            }
+        }
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:59]:
```suggestion
        final List<QueryBuilder> queries = new ArrayList<>();
        QueryBuilder filter = null;
        String queryName = null;

        for (QueryContainer queryContainer : hybridQueryProto.getQueriesList()) {
            QueryBuilder queryBuilder = parseInnerQueryBuilder(queryContainer, registry);
            if (queryBuilder != null) {
                queries.add(queryBuilder);
            }
        }
        
        if (queries.size() > HybridQueryBuilder.MAX_NUMBER_OF_SUB_QUERIES) {
                throw new IllegalArgumentException(
                    String.format(Locale.ROOT, HybridQueryBuilder.ERROR_MSG_MAX_QUERIES_EXCEEDED, HybridQueryBuilder.NAME)
                );
       }
```

Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:59]: Wouldn't we want to fail fast instead of adding it to the array if it will be rejected later on? 

Also the REST side follows the same logic to throw exception before adding the query to the array - I think we should keep it as is to maintain parity with the REST side, wdyt? 
https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java#L223-L231

**Comment by karenyrx** [comment]
Hi @vibrantvarun @martin-gaievski could you help merge this PR if all comments are addressed? Thankd!

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+        boolean hasInnerHits = false;
+        for (QueryBuilder query : queries) {
+            if (filter == null) {
+                compoundQueryBuilder.add(query);
+            } else {
+                compoundQueryBuilder.add(query.filter(filter));
+            }
+
+            if (hasInnerHits == false) {
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:103]: The following code tries to extract innerhits when hasInnerHits is false. Why?

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 8 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:59]: I was going behind the readability aspect of it. Anyways I am fine either.

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 9 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:103]: It matches the REST side 
```
        boolean hasInnerHits = false;
        for (QueryBuilder query : queries) {
            if (filter == null) {
                compoundQueryBuilder.add(query);
            } else {
                compoundQueryBuilder.add(query.filter(filter));
            }

            // Check if children have inner hits for stats
            if (hasInnerHits == false) {
                Map<String, InnerHitContextBuilder> innerHits = new HashMap<>();
                InnerHitContextBuilder.extractInnerHits(query, innerHits);
                hasInnerHits = innerHits.isEmpty() == false;
            }
        }

```
That is a question for neural-search maintainers :)

Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:103]: https://github.com/opensearch-project/neural-search/blame/main/src/main/java/org/opensearch/neuralsearch/query/HybridQueryBuilder.java#L277C9-L277C16

File: src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java (hunk 8 above)
Comment (karenyrx) [reply src/main/java/org/opensearch/neuralsearch/grpc/proto/request/search/query/HybridQueryBuilderProtoUtils.java:59]: Got it, let's keep it as is to keep the gRPC/REST sides in sync then
(Though FWIW, I think the old code was quite readable actually :))

**Comment by karenyrx** [comment]
Adding integration test in https://github.com/opensearch-project/neural-search/pull/1734
